File: app/src/main/java/com/stupidtree/hichat/ui/relationevent/RelationEventActivity.java
package com.stupidtree.hichat.ui.relationevent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Context;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.stupidtree.hichat.R;
import com.stupidtree.hichat.data.model.RelationEvent;
import com.stupidtree.hichat.ui.base.BaseActivity;
import com.stupidtree.hichat.ui.base.BaseListAdapter;
import com.stupidtree.hichat.ui.base.BaseViewHolder;
import com.stupidtree.hichat.ui.base.DataState;
import com.stupidtree.hichat.ui.widgets.PopUpCheckableList;
import com.stupidtree.hichat.utils.ActivityUtils;
import com.stupidtree.hichat.utils.ImageUtils;
import com.stupidtree.hichat.utils.TextUtils;

import java.lang.reflect.Array;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Objects;

import butterknife.BindView;

/**
 * 好友关系页面
 */
public class RelationEventActivity extends BaseActivity<RelationEventViewModel> {

    /**
     * View绑定区
     */
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.list)
    RecyclerView list;

    /**
     * 适配器
     */
    RListAdapter listAdapter;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_relation_event;
    }

    @Override
    protected Class<RelationEventViewModel> getViewModelClass() {
        return RelationEventViewModel.class;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setWindowParams(true,true,false);
        setToolbarActionBack(toolbar);
    }

    @Override
    protected void initViews() {
        listAdapter = new RListAdapter(this,new LinkedList<>());
        list.setAdapter(listAdapter);
        list.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        listAdapter.setOnButtonClickListener((data, card, position) -> {
            new PopUpCheckableList<RelationEvent.ACTION>().setTitle(R.string.select_action)
                    .setListData(
                            Arrays.asList(getString(R.string.prompt_accept),getString(R.string.prompt_reject)),
                            Arrays.asList(RelationEvent.ACTION.ACCEPT, RelationEvent.ACTION.REJECT))
                    .setOnConfirmListener((title, key) -> viewModel.responseFriendRequest(data.getId(), key))
            .show(getSupportFragmentManager(),"select");

        });
        listAdapter.setOnItemClickListener((data, card, position) -> ActivityUtils.startProfileActivity(getThis(),data.getOtherId()));

        viewModel.getListData().observe(this, listDataState -> {
            if(listDataState.getState()== DataState.STATE.SUCCESS){
                listAdapter.notifyItemChangedSmooth(listDataState.getData(), (a, b) -> !Objects.equals(a, b));
            }
        });
        viewModel.getResponseResult().observe(this, dataState -> {
            if(dataState.getState()== DataState.STATE.SUCCESS){
                Toast.makeText(getThis(),R.string.make_friends_success,Toast.LENGTH_SHORT).show();
            }
            viewModel.startRefresh();
        });


        viewModel.getMarkReadResult().observe(this, dataState -> {
        });
    }


    @Override
    protected void onResume() {
        super.onResume();
        viewModel.startRefresh();
        viewModel.startMarkRead();
    }

    class RListAdapter extends BaseListAdapter<RelationEvent, RListAdapter.RHolder>{

        OnItemClickListener<RelationEvent> mOnButtonClickListener;

        public void setOnButtonClickListener(OnItemClickListener<RelationEvent> mOnButtonClickListener) {
            this.mOnButtonClickListener = mOnButtonClickListener;
        }

        public RListAdapter(Context mContext, List<RelationEvent> mBeans) {
            super(mContext, mBeans);
        }

        @Override
        protected int getLayoutId(int viewType) {
            return R.layout.activity_relation_event_list_item;
        }

        @Override
        public RHolder createViewHolder(View v, int viewType) {
            return new RHolder(v);
        }

        @Override
        protected void bindHolder(@NonNull RHolder holder, @Nullable RelationEvent data, int position) {
            if(data!=null){
                holder.nickname.setText(data.getOtherNickname());
                holder.time.setText(TextUtils.getChatTimeText(mContext,data.getCreatedAt()));
                ImageUtils.loadAvatarNoCacheInto(mContext,data.getOtherAvatar(),holder.avatar);
                if(mOnItemClickListener!=null){
                    holder.item.setOnClickListener(view -> mOnItemClickListener.onItemClick(data,view,position));
                }
                if(data.isUnread()){
                    holder.unread.setVisibility(View.VISIBLE);
                }else{
                    holder.unread.setVisibility(View.GONE);
                }

                if(Objects.equals(data.getUserId(),viewModel.getLocalUserId())){ //我发出的
                    holder.accept.setVisibility(View.GONE);
                    holder.message.setVisibility(View.VISIBLE);
                    holder.icon.setImageResource(R.drawable.ic_sent);
                    if(data.getState()== RelationEvent.STATE.REQUESTING){
                        holder.message.setText(R.string.relation_requesting);
                    }else if(data.getState() == RelationEvent.STATE.REJECTED){
                        holder.message.setText(R.string.relation_rejected);
                    }else if(data.getState()== RelationEvent.STATE.ACCEPTED){
                        holder.message.setText(R.string.relation_accepted);
                    }else if(data.getState()== RelationEvent.STATE.DELETE){
                        holder.message.setText(R.string.relation_you_deleted);
                    }
                }else{ //我收到的
                    holder.icon.setImageResource(R.drawable.ic_received);
                    holder.accept.setVisibility(View.GONE);
                    holder.message.setVisibility(View.VISIBLE);
                    if(data.getState()== RelationEvent.STATE.DELETE){
                        holder.message.setText(R.string.relation_delete);
                    }else if(data.getState()== RelationEvent.STATE.ACCEPTED) {
                        holder.message.setText(R.string.relation_accepted);
                    } else if(data.getState() == RelationEvent.STATE.REJECTED){
                        holder.message.setText(R.string.relation_rejected);
                    }else if(data.getState()== RelationEvent.STATE.REQUESTING){
                        holder.accept.setVisibility(View.VISIBLE);
                        holder.message.setVisibility(View.GONE);
                        if(mOnButtonClickListener!=null){
                            holder.accept.setOnClickListener(view -> mOnButtonClickListener.onItemClick(data,view,position));
                        }
                    }
                }
                if(data.getState()== RelationEvent.STATE.REQUESTING){
                    holder.icon.setImageTintList(ColorStateList.valueOf(getColor(R.color.colorPrimary)));
                    holder.icon.setBackgroundResource(R.drawable.element_round_primary);
                }else{
                    holder.icon.setImageTintList(ColorStateList.valueOf(getColor(android.R.color.darker_gray)));
                    holder.icon.setBackgroundResource(R.drawable.element_round_grey);
                }
            }
        }

        class RHolder extends RecyclerView.ViewHolder {
            View item;
            ImageView avatar;
            TextView nickname;
            ImageView icon;
            TextView message;
            TextView time;
            View accept;
            ImageView unread;
            public RHolder(@NonNull View itemView) {
                super(itemView);
                item = itemView.findViewById(R.id.item);
                accept = itemView.findViewById(R.id.accept);
                avatar = itemView.findViewById(R.id.avatar);
                nickname = itemView.findViewById(R.id.nickname);
                icon = itemView.findViewById(R.id.icon);
                message = itemView.findViewById(R.id.message);
                time = itemView.findViewById(R.id.time);
                unread = itemView.findViewById(R.id.unread);
            }
        }
    }

}

File: app/src/main/java/com/stupidtree/cloudliter/ui/relationevent/RelationEventActivity.kt
package com.stupidtree.cloudliter.ui.relationevent

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.viewbinding.ViewBinding
import com.stupidtree.cloudliter.R
import com.stupidtree.cloudliter.data.model.RelationEvent
import com.stupidtree.cloudliter.data.model.RelationEvent.ACTION
import com.stupidtree.cloudliter.databinding.ActivityRelationEventBinding
import com.stupidtree.cloudliter.databinding.ActivityRelationEventListItemBinding
import com.stupidtree.cloudliter.ui.base.BaseActivity
import com.stupidtree.cloudliter.ui.base.BaseListAdapter
import com.stupidtree.cloudliter.ui.base.BaseListAdapter.RefreshJudge
import com.stupidtree.cloudliter.ui.base.BaseViewHolder
import com.stupidtree.cloudliter.ui.base.DataState
import com.stupidtree.cloudliter.ui.relationevent.RelationEventActivity.RListAdapter.RHolder
import com.stupidtree.cloudliter.ui.widgets.PopUpCheckableList
import com.stupidtree.cloudliter.utils.ActivityUtils
import com.stupidtree.cloudliter.utils.ImageUtils
import com.stupidtree.cloudliter.utils.TextUtils
import java.util.*

/**
 * 好友关系页面
 */
@SuppressLint("NonConstantResourceId")
class RelationEventActivity : BaseActivity<RelationEventViewModel,ActivityRelationEventBinding>() {
    /**
     * 适配器
     */
    private lateinit var listAdapter: RListAdapter

    override fun getViewModelClass(): Class<RelationEventViewModel> {
        return RelationEventViewModel::class.java
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setWindowParams(true, darkColor = true, navi = false)
        setToolbarActionBack(binding.toolbar)
    }

    override fun initViews() {
        listAdapter = RListAdapter(this, LinkedList())
        binding.list.adapter = listAdapter
        binding.list.layoutManager = LinearLayoutManager(applicationContext)
        listAdapter.setOnButtonClickListener(
                object : BaseListAdapter.OnItemClickListener<RelationEvent> {
                    override fun onItemClick(data: RelationEvent, card: View?, position: Int) {
                        PopUpCheckableList<ACTION>().setTitle(getString(R.string.select_action))
                                .setListData(
                                        listOf(getString(R.string.prompt_accept), getString(R.string.prompt_reject)),
                                        listOf(ACTION.ACCEPT, ACTION.REJECT))
                                .setOnConfirmListener (object:PopUpCheckableList.OnConfirmListener<ACTION> {
                                    override fun OnConfirm(title: String?, key: ACTION) {
                                        key.let { data.id?.let { it1 -> viewModel.responseFriendRequest(it1, it) } }
                                    }
                                })
                                .show(supportFragmentManager, "select")
                    }
                })
        listAdapter.setOnItemClickListener(object : BaseListAdapter.OnItemClickListener<RelationEvent> {
            override fun onItemClick(data: RelationEvent, card: View?, position: Int) {
                data.otherId?.let { ActivityUtils.startProfileActivity(getThis(), it) }
            }
        })
        viewModel.listData?.observe(this, Observer { listDataState: DataState<List<RelationEvent>?> ->
            if (listDataState.state === DataState.STATE.SUCCESS) {
                listDataState.data?.let {
                    listAdapter.notifyItemChangedSmooth(it, object : RefreshJudge<RelationEvent> {
                        override fun judge(oldData: RelationEvent, newData: RelationEvent): Boolean {
                            return oldData != newData
                        }
                    })
                }
            }
        })
        viewModel.responseResult?.observe(this, Observer { dataState: DataState<*> ->
            if (dataState.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.make_friends_success, Toast.LENGTH_SHORT).show()
            }
            viewModel.startRefresh()
        })
        viewModel.markReadResult?.observe(this, { })
    }

    override fun onResume() {
        super.onResume()
        viewModel.startRefresh()
        viewModel.startMarkRead()
    }

    @SuppressLint("ParcelCreator")
    private inner class RListAdapter(mContext: Context, mBeans: MutableList<RelationEvent>) : BaseListAdapter<RelationEvent, RHolder>(mContext, mBeans) {
        var mOnButtonClickListener: OnItemClickListener<RelationEvent>? = null
        fun setOnButtonClickListener(mOnButtonClickListener: OnItemClickListener<RelationEvent>?) {
            this.mOnButtonClickListener = mOnButtonClickListener
        }



        override fun bindHolder(holder: RHolder, data: RelationEvent?, position: Int) {
            if (data != null) {
                holder.binding.nickname.text = data.otherNickname
                holder.binding.time.text = TextUtils.getChatTimeText(mContext, data.createdAt)
                ImageUtils.loadAvatarNoCacheInto(mContext, data.otherAvatar, holder.binding.avatar)
                if (mOnItemClickListener != null) {
                    holder.binding.item.setOnClickListener { view: View? -> mOnItemClickListener!!.onItemClick(data, view, position) }
                }
                if (data.isUnread) {
                    holder.binding.unread.visibility = View.VISIBLE
                } else {
                    holder.binding.unread.visibility = View.GONE
                }
                if (data.userId == viewModel.localUserId) { //我发出的
                    holder.binding.accept.visibility = View.GONE
                    holder.binding.message.visibility = View.VISIBLE
                    holder.binding.icon.setImageResource(R.drawable.ic_sent)
                    if (data.state == RelationEvent.STATE.REQUESTING) {
                        holder.binding.message.setText(R.string.relation_requesting)
                    } else if (data.state == RelationEvent.STATE.REJECTED) {
                        holder.binding.message.setText(R.string.relation_rejected)
                    } else if (data.state == RelationEvent.STATE.ACCEPTED) {
                        holder.binding.message.setText(R.string.relation_accepted)
                    } else if (data.state == RelationEvent.STATE.DELETE) {
                        holder.binding.message.setText(R.string.relation_you_deleted)
                    }
                } else { //我收到的
                    holder.binding.icon.setImageResource(R.drawable.ic_received)
                    holder.binding.accept.visibility = View.GONE
                    holder.binding.message.visibility = View.VISIBLE
                    if (data.state == RelationEvent.STATE.DELETE) {
                        holder.binding.message.setText(R.string.relation_delete)
                    } else if (data.state == RelationEvent.STATE.ACCEPTED) {
                        holder.binding.message.setText(R.string.relation_accepted)
                    } else if (data.state == RelationEvent.STATE.REJECTED) {
                        holder.binding.message.setText(R.string.relation_rejected)
                    } else if (data.state == RelationEvent.STATE.REQUESTING) {
                        holder.binding.accept.visibility = View.VISIBLE
                        holder.binding.message.visibility = View.GONE
                        if (mOnButtonClickListener != null) {
                            holder.binding.accept.setOnClickListener { view: View? -> mOnButtonClickListener!!.onItemClick(data, view, position) }
                        }
                    }
                }
                if (data.state == RelationEvent.STATE.REQUESTING) {
                    holder.binding.icon.imageTintList = ColorStateList.valueOf(ContextCompat.getColor(getThis(), R.color.colorPrimary))
                    holder.binding.icon.setBackgroundResource(R.drawable.element_round_primary)
                } else {
                    holder.binding.icon.imageTintList = ColorStateList.valueOf(ContextCompat.getColor(getThis(), android.R.color.darker_gray))
                    holder.binding.icon.setBackgroundResource(R.drawable.element_round_grey)
                }
            }
        }

        inner class RHolder(itemView: ActivityRelationEventListItemBinding) : BaseViewHolder<ActivityRelationEventListItemBinding>(itemView)

        override fun getViewBinding(parent: ViewGroup, viewType: Int): ViewBinding {
            return ActivityRelationEventListItemBinding.inflate(mInflater,parent,false)
        }

        override fun createViewHolder(viewBinding: ViewBinding, viewType: Int): RHolder {
            return RHolder(viewBinding as ActivityRelationEventListItemBinding)
        }
    }

    override fun initViewBinding(): ActivityRelationEventBinding {
        return ActivityRelationEventBinding.inflate(layoutInflater)
    }
}